src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala

```diff
@@ -390,8 +390,13 @@ trait ContextErrors {
       }
 
       //typedNew
-      def IsAbstractError(tree: Tree, sym: Symbol) = {
-        issueNormalTypeError(tree, sym + " is abstract; cannot be instantiated")
+      def IsAbstractError(tree: Tree, sym: Symbol, context: Context) = {
+        issueNormalTypeError(tree, sym + s" is ${if (sym.isTrait) "trait" else "abstract"}; cannot be ${if(context.inAnnotation) "annotated" else "instantiated"}")
```

Comment:
"cannot be annotated" has the wrong subject, in `@Foo X`, "X" is annotated, not "Foo". I'd keep the original message ("Foo cannot be instantiated") or be more explicit ("cannot be used as an annotation"). I'm also not sure if "Foo is trait" is an improvement, I find "Foo is abstract" a better explanation.

Reply:
I'd vote for the "more explicit" one.